analysis/src/ProcessCmt.ml

```diff
@@ -793,3 +793,77 @@ let fileForModule moduleName ~package =
   | None ->
     Log.log ("No path for module " ^ moduleName);
     None
+
+(* Collect top-level exception constructors from typedtree/CMT file. *)
+let exceptionsForCmt ~cmt : (string * bool) list =
+  match Shared.tryReadCmt cmt with
+  | None -> []
+  | Some infos ->
+    let by_name : (string, bool) Hashtbl.t = Hashtbl.create 16 in
+    let add_ext (ext : Typedtree.extension_constructor) : unit =
+      let name = ext.ext_name.txt in
+      let hasArgs =
+        match ext.ext_kind with
+        | Text_decl (Cstr_tuple args, _ret) -> args <> []
+        | Text_decl (Cstr_record fields, _ret) -> fields <> []
+        | Text_rebind _ -> true
+      in
+      let prev =
+        match Hashtbl.find_opt by_name name with
+        | Some b -> b
+        | None -> false
+      in
+      Hashtbl.replace by_name name (prev || hasArgs)
+    in
+    (* Only collect top-level exception declarations (Tstr_exception/Tsig_exception).
+     Avoid picking up exceptions from Texp_letexception by tracking context. *)
+    let in_toplevel_exception = ref false in
+    let module Iter = TypedtreeIter.MakeIterator (struct
+      include TypedtreeIter.DefaultIteratorArgument
```

Comment:
Do we use the functor based approach elsewhere in the code base? Or is there no way for the typed tree to just define a mapper like we do for the AST mapper? I don't remember how this works exactly.